### ISSUE TITLE
Window appears too tall when there is a wrapped item at top level (GridWrap / Label.Wrapping etc)

### ISSUE THREAD

**Issue by Klaus-Tockloth** [fyne-io/fyne#4097]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

If Window.Resize() is used after Window.SetContent(), the height setting seems to be ignored.

### How to reproduce

Build and run example code.

### Screenshots

<img width="546" alt="Bildschirmfoto 2023-07-25 um 17 23 25" src="https://github.com/fyne-io/fyne/assets/5493128/d512372f-6a8e-42b3-924d-45fad5293735">

### Example code

```
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/widget"
)

func main() {
	application := app.New()

	initProgressContainer1 := createInitializationProgressContainer()
	wMain := application.NewWindow("wMain")
	wMain.SetMaster()
	wMain.Resize(fyne.NewSize(200, 50)) // width, pseudo height
	wMain.SetContent(initProgressContainer1)
	wMain.Show()

	initProgressContainer2 := createInitializationProgressContainer()
	wSecond := application.NewWindow("wSecond")
	// wSecond.Resize(fyne.NewSize(300, 50)) // OK
	wSecond.SetContent(initProgressContainer2)
	wSecond.Resize(fyne.NewSize(300, 50)) // NOK
	wSecond.Show()

	application.Run()
}

func createInitializationProgressContainer() fyne.CanvasObject {
	messageLabel1 := widget.NewLabelWithStyle("Message:", fyne.TextAlignLeading, fyne.TextStyle{Bold: true, Italic: true})
	messageLabel1.Wrapping = fyne.TextWrapWord
	messageLabel2 := widget.NewLabel("Initialization ... please wait.")
	messageLabel2.Wrapping = fyne.TextWrapWord

	return container.New(
		layout.NewVBoxLayout(),
		widget.NewSeparator(),
		messageLabel1,
		messageLabel2,
		widget.NewSeparator(),
	)
}
```

### Fyne version

2.3.5

### Go compiler version

1.20

### Operating system and version

macOS 13.4.1

**Comment by andydotxyz** [comment]
This is an issue with wrapping and macOS windows - it will not always respect the fact that the minimum window size has been reduced (calculating sizes with wrapping turned on is a 2-pass process).

I think this is a duplicate of #3252

**Comment by andydotxyz** [comment]
(p.s. don't set word wrap on a label with only 1 word - it is adding layout processing that is expensive when it will never add anything)

**Comment by Klaus-Tockloth** [comment]
Windows: The height of both windows are incorrect.

![init-windows](https://github.com/fyne-io/fyne/assets/5493128/17b8459c-b704-4e09-ae4e-a858f5ff187d)

**Comment by Klaus-Tockloth** [comment]
Linux: The The height of both windows is incorrect.

![init-linux](https://github.com/fyne-io/fyne/assets/5493128/605d7b20-2ce3-4e3e-b8b4-afca83cd61b3)

(PS: Not related to this issue: Each application on my Ubuntu uses the dark theme as default. How can this be controlled?

**Comment by andydotxyz** [comment]
> (PS: Not related to this issue: Each application on my Ubuntu uses the dark theme as default. How can this be controlled?)

On develop it should pick up your KDE/Gnome preference. Until then you can set FYNE_THEME=dark or use fyne_settings app.

**Comment by Klaus-Tockloth** [comment]
I have made an additional test on Citrix (Windows 10 Enterprise) with the OpenGL emulation "MESA3D". Same result as on standard Windows.

![image001](https://github.com/fyne-io/fyne/assets/5493128/76b39b64-c360-4524-8297-d89c958b420b)

**Comment by Klaus-Tockloth** [comment]
After a lot of testing ... let me draw a conclusion:

Window.Resize() is not properly working on 4 different OpenGL implementations:
- Windows (OpenGL driver)
- Citrix (virtual Windows, OpenGL emulation) 
- Ubuntu (OpenGL driver)
- macOS (OpenGL driver)

The behavior on macOS is a bit different but also incorrect.

For me this looks like a bug in Fyne and not in OpenGL. Do you agree with that?

Also possible (but unlikely): Misuse of Window.Resize() within the test program (see above).

PS: My use case is something like an "one page application" with a fix width and variable height. This means, that a properly working Window.Resize() function is very important for me.

**Comment by andydotxyz** [comment]
> For me this looks like a bug in Fyne and not in OpenGL. Do you agree with that?

Apologies I did not mean to imply this was a bug with OpenGL - the size of a window is not related to how we paint it.
As outlined above it seems to relate to the multi-pass method for calculating height required for a wrapped item in the screen and some OS don't allow the window to be sized smaller after we have made it larger. (Pass 1 knows how long the wrap would be required if narrow, we then find out how wide it is and re-calculate the height required. Seems like this should happen all before it shows but that isn't working - that seems like the bug.

**Comment by Jacalz** [comment]
> On develop it should pick up your KDE/Gnome preference. Until then you can set FYNE_THEME=dark or use fyne_settings app.

That support is already available in the stable releases. It might just be a case of the Ubuntu version being too old to support the new light/dark theme preference protocol.

**Comment by Klaus-Tockloth** [comment]
I have tested the issue again with Fyne v2.5.2. Unfortunately, the problem persists and can be reproduced (see screenshot under Windows).

<img width="292" alt="Bildschirmfoto 2024-11-07 um 09 31 33" src="https://github.com/user-attachments/assets/a6a5861e-7b18-4cc1-b833-66a5086f56b7">

Question: Could someone take a look at the situation?

**Comment by andydotxyz** [comment]
Not to be one for workarounds but I can't help but think that your UI looks like it will have a lot of messages. Wouldn't putting them in a Scroll container or List be a good thing to do?

If so the window size issue will go away because the wrapping will not be pushing the window to be larger during initialisation.

**Comment by Klaus-Tockloth** [comment]
Hmm, the code to reproduce the behavior "miscalculation of window height" is very simple and only has one message. Shouldn't it be marked as 'bug'?

**Comment by Klaus-Tockloth** [comment]
It seems that the issue has something to do with text wrapping ("fyne.TextWrapWord").

Without any text wrapping I get correct window sizes:
<img width="320" alt="Bildschirmfoto 2024-11-18 um 08 55 50" src="https://github.com/user-attachments/assets/a0eea3d9-a27b-4d99-8cf1-45827d19354e">

With text wrapping "fyne.TextWrapWord" I get incorrect window sizes:
<img width="328" alt="Bildschirmfoto 2024-11-18 um 08 56 41" src="https://github.com/user-attachments/assets/b0041a45-b8bc-464d-8f8b-199ec1b3d5d1">

**Comment by andydotxyz** [comment]
> It seems that the issue has something to do with text wrapping ("fyne.TextWrapWord").

Yes, that is what I wrote on my first reply a year ago.
And if the wrapped content is put inside a vertical scroller the issue goes away because it no longer impacts window size.

**Comment by Klaus-Tockloth** [comment]
Yes, the defect is caused by word wrapping of texts. I hope this information helps to find and correct the defect.

**Comment by andydotxyz** [comment]
Marking as blocker because this issue got worse recently, now impacting all OS.

**Comment by dweymouth** [comment]
Confirmed fixed on Mac and Linux, closing this as it is likely resolved for Windows as well, but feel free to reopen if it repros on fyne develop.
